File: HCsbLib/HCsbLib/HTLib2/HMath.Geometry/Geometry.DerivativeOfTriangleRadius.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace HTLib2
{
	public partial class Geometry
	{
        public static double DerivativeOfTriangleRadius(Vector p1, Vector p2, Vector p3, Vector dp1, Vector dp2, Vector dp3)
        {
            /// r = |p1-p2|*|p2-p3|*|p3-p1| / (2 area(p1,p2,p3))
            ///   = |p1-p2|*|p2-p3|*|p3-p1| / (2 |(p1-p2)x(p2-p3)|)
            /// r2 = ( |p1-p2|*|p2-p3|*|p3-p1| / (2 area(p1,p2,p3)) )^2
            ///    = |p1-p2|^2 * |p2-p3|^2 * |p3-p1|^2 / (4 |(p1-p2)x(p2-p3)|^2)
            /// (r+dr*t)  = R  = |(p1+dp1*t)-(p2+dp2*t)|*|(p2+dp2*t)-(p3+dp3*t)|*|(p3+dp3*t)-(p1+dp1*t)| / (2 area(p1+dp1*t,p2+dp2*t,p3+dp3*t))
            /// (r+dr*t)2 = R2 = { |(p1+dp1*t)-(p2+dp2*t)|*|(p2+dp2*t)-(p3+dp3*t)|*|(p3+dp3*t)-(p1+dp1*t)| / (2 area(p1+dp1*t,p2+dp2*t,p3+dp3*t)) }^2
            ///   
            /// d(r+dr*t)/dt = dR_dt
            ///              = d(r+dr*t)/d((r+dr*t)^2) * d((r+dr*t)^2)/dt
            ///              = dR_dR2 * dR2_dt
            /// d(r+dr*t)/d((r+dr*t)^2) = dR_dR2
            ///                         = 1/(2*r)
            /// d((r+dr*t)^2)/dt = dR2_dt
            ///                = + (   Dotp31dp31 * P12dist2 * P23dist2
            ///                      + Dotp23dp31 * P12dist2 * P31dist2
            ///                      + Dotp12dp12 * P23dist2 * P31dist2
            ///                    )/(2 * Area2)
            ///                  - ( DotCroP12P23CroP12Dp23CroDp12P23 * P12dist2 * P23dist2 * P31dist2)/(2 * Area4)
            ///
            /// P12dist2 := ((p1x-p2x)^2+(p1y-p2y)^2+(p1z-p2z)^2) = (p1-p2).dist2
            /// P23dist2 := ((p2x-p3x)^2+(p2y-p3y)^2+(p2z-p3z)^2) = (p2-p3).dist2
            /// P31dist2 := ((p1x-p3x)^2+(p1y-p3y)^2+(p1z-p3z)^2) = (p1-p3).dist2
            /// Dotp12dp12 := ((dp1x-dp2x)*(p1x-p2x) + (dp1y-dp2y)*(p1y-p2y) + (dp1z-dp2z)*(p1z-p2z))
            ///             = [dp1x-dp2x, dp1y-dp2y, dp1z-dp2z] . [p1x-p2x, p1y-p2y, p1z-p2z]
            ///             = (dp1-dp2).(p1-p2)
            /// Dotp23dp31 := ((dp2x-dp3x)*(p2x-p3x) + (dp2y-dp3y)*(p2y-p3y) + (dp2z-dp3z)*(p2z-p3z))
            ///             = (dp2-dp3).(p2-p3)
            /// Dotp31dp31 := ((dp1x-dp3x)*(p1x-p3x) + (dp1y-dp3y)*(p1y-p3y) + (dp1z-dp3z)*(p1z-p3z))
            ///             = (dp1-dp3).(p1-p3)
            /// DotCroP12P23CroP12Dp23CroDp12P23 := Dot[Cross[p1-p2,p2-p3],(Cross[p1-p2,dp2-dp3]+Cross[dp1-dp2,p2-p3])]
            ///                                   = [(p1-p2)x(p2-p3)] . [(p1-p2)x(dp2-dp3) + (dp1-dp2)x(p2-p3)]
                
            if(HDebug.Selftest())
                #region selftest
            {
                Vector lp1 = new double[] { 1, 0, 0 };
                Vector lp2 = new double[] { 0, 1, 0 };
                Vector lp3 = new double[] { 0, 0, 1 };
                Vector ldp1 = new double[] { 0.1, 0.1, 0.1 };
                Vector ldp2 = new double[] { 1, 2, 3 };
                Vector ldp3 = new double[] { -1, 0, -0.1 };

                double r0 = Geometry.RadiusOfTriangle(lp1, lp2, lp3);
                double dr0 = DerivativeOfTriangleRadius(lp1, lp2, lp3, ldp1, ldp2, ldp3);
                List<Tuple<double, double, double, double>> drx = new List<Tuple<double, double, double, double>>();
                foreach(double dt in new double[] { 0.1, 0.01, 0.001, 0.0001, 0.00001, 0.000001, 0.0000001 })
                {
                    double t00 = -dt; double r00 = Geometry.RadiusOfTriangle(lp1+t00*ldp1, lp2+t00*ldp2, lp3+t00*ldp3);
                    double t01 = +dt; double r01 = Geometry.RadiusOfTriangle(lp1+t01*ldp1, lp2+t01*ldp2, lp3+t01*ldp3);
                    double dr1 = (r01    -r00)/(t01-t00);
                    double dr2 = (r01*r01-r00*r00)/(t01-t00);

                    drx.Add(new Tuple<double, double, double, double>(dt, dr1, dr2, (0.5/r0)*dr2));
                }
                double diff = dr0-drx.Last().Item2;
                HDebug.AssertTolerance(0.00000001, diff);
            }
                #endregion

            double P12dist2 = (p1-p2).Dist2;
            double P23dist2 = (p2-p3).Dist2;
            double P31dist2 = (p1-p3).Dist2;
            double Dotp12dp12 = LinAlg.VtV(dp1-dp2, p1-p2);
            double Dotp23dp31 = LinAlg.VtV(dp2-dp3, p2-p3);
            double Dotp31dp31 = LinAlg.VtV(dp1-dp3, p1-p3);
            double DotCroP12P23CroP12Dp23CroDp12P23 = LinAlg.VtV( LinAlg.CrossProd(p1-p2,  p2- p3)
                                                                , LinAlg.CrossProd(p1-p2, dp2-dp3)+LinAlg.CrossProd(dp1-dp2, p2-p3)
                                                                );
            double Area2 = LinAlg.CrossProd(p1-p2, p2-p3).Dist2;
            double Area4 = Area2*Area2;
            double Rad2 = P12dist2 * P23dist2 * P31dist2 / (4*Area2);
            double Rad  = Math.Sqrt(Rad2);
            HDebug.AssertTolerance(0.00000001, Rad-Geometry.RadiusOfTriangle(p1, p2, p3));

            double t = 1;
            double dR_dR2 = 0.5/Rad;
            double dR2_dt = ((t*Dotp31dp31)*P12dist2*P23dist2 + (t*Dotp23dp31)*P12dist2*P31dist2 + (t*Dotp12dp12)*P23dist2*P31dist2)/(2 * Area2)
                            - ((t*DotCroP12P23CroP12Dp23CroDp12P23) * P12dist2 * P23dist2 * P31dist2)/(2 * Area4);
            double dR_dt = dR_dR2 * dR2_dt;
            return dR_dt;
        }
    }
}

File: HCsbLib/HCsbLib/HTLib2/HMath.Geometry/Geometry.DerivativeOfTriangleRadius.m
(* ::Package:: *)

p1={p1x,p1y,p1z};
p2={p2x,p2y,p2z};
p3={p3x,p3y,p3z};
dp1={dp1x,dp1y,dp1z};
dp2={dp2x,dp2y,dp2z};
dp3={dp3x,dp3y,dp3z};

(*
http://www.mathalino.com/reviewer/derivation-of-formulas/derivation-of-formula-for-radius-of-circumcircle
Radius R of triangle is
   R = abc / 4A,
where a,b,c are the lengths of sides of triangle, and A is the area of triangle.

http://en.wikipedia.org/wiki/Triangle#Using_vectors
Area of triangle is
  0.5 |AB x AC|,
where A,B,C are the location of points, AB is the vector from A to B, and 'x' is cross-product.
*)
R2[p1_,p2_,p3_] := ((p1-p2).(p1-p2)) * ((p2-p3).(p2-p3)) * ((p3-p1).(p3-p1)) / (4*Cross[p1-p2, p2-p3].Cross[p1-p2, p2-p3]);

R2a[p1_,p2_,p3_]:= ((p1-p2).(p1-p2)) * ((p2-p3).(p2-p3)) * ((p3-p1).(p3-p1));
R2b[p1_,p2_,p3_]:=                                                               (4*Cross[p1-p2, p2-p3].Cross[p1-p2, p2-p3]);
R2[p1_,p2_,p3_] := R2a[p1,p2,p3] / R2b[p1,p2,p3];

R2at[p1_,p2_,p3_,dp1_,dp2_,dp3_,t_] := R2a[p1+dp1*t,p2+dp2*t,p3+dp3*t];
R2bt[p1_,p2_,p3_,dp1_,dp2_,dp3_,t_] := R2b[p1+dp1*t,p2+dp2*t,p3+dp3*t];
R2t [p1_,p2_,p3_,dp1_,dp2_,dp3_,t_] := R2 [p1+dp1*t,p2+dp2*t,p3+dp3*t];

dR2at[p1_,p2_,p3_,dp1_,dp2_,dp3_,t_] := D[R2at[p1,p2,p3,dp1,dp2,dp3,t],t];
dR2bt[p1_,p2_,p3_,dp1_,dp2_,dp3_,t_] := D[R2bt[p1,p2,p3,dp1,dp2,dp3,t],t];
dR2t [p1_,p2_,p3_,dp1_,dp2_,dp3_,t_] := D[R2t [p1,p2,p3,dp1,dp2,dp3,t],t];


f[a_,b_,c_,da_,db_,dc_] := ((da-db).(a-b)) * ((a-c).(a-c)) * ((b-c).(b-c));
ff[a_,b_,c_,da_,db_,dc_] := f[a,b,c,da,db,dc]+f[c,a,b,dc,da,db]+f[b,c,a,db,dc,da];
(* ff[...] = f(a,b,c) + f(c,a,b) + f(b,c,a) *)
dR2at0[p1_,p2_,p3_,dp1_,dp2_,dp3_] := Limit[dR2at[p1,p2,p3,dp1,dp2,dp3,t],t->0];
dR2at0[p1,p2,p3,dp1,dp2,dp3] - 2*ff[p1,p2,p3,dp1,dp2,dp3] //Simplify
(******************************* should be 0 *******************************)


gg[a_,b_,c_,da_,db_,dc_] := Cross[a-b,b-c] . (Cross[a-b, db-dc] + Cross[da-db, b-c]);
(* g[...] = (Pab x Pbc)^tr (Pab x Wbc + Wab x Pbc)
      Wbc = Wa - Wb = da - db
      Pab = Pa - Pb = a - b
      a x b = cross[a,b]
*)
dR2bt0[p1_,p2_,p3_,dp1_,dp2_,dp3_] := Limit[dR2bt[p1,p2,p3,dp1,dp2,dp3,t],t->0];
dR2bt0[p1,p2,p3,dp1,dp2,dp3] - 2*(4*gg[p1,p2,p3,dp1,dp2,dp3]) //Simplify
(******************************* should be 0 *******************************)


( dR2at0[p1,p2,p3,dp1,dp2,dp3]*R2bt[p1,p2,p3,dp1,dp2,dp3,0]
- Limit[dR2bt[p1,p2,p3,dp1,dp2,dp3,tt],tt->0]*R2at[p1,p2,p3,dp1,dp2,dp3,0]
) - 2*(
    ff[p1,p2,p3,dp1,dp2,dp3]*R2bt[p1,p2,p3,dp1,dp2,dp3,0]
- 4*gg[p1,p2,p3,dp1,dp2,dp3]*R2at[p1,p2,p3,dp1,dp2,dp3,0]
) //Simplify
(******************************* should be 0 *******************************)

dABdBAt[p1_,p2_,p3_,dp1_,dp2_,dp3_,t_] := (Limit[dR2at[p1,p2,p3,dp1,dp2,dp3,tt],tt->0]*R2bt[p1,p2,p3,dp1,dp2,dp3,t]
                                         - R2at[p1,p2,p3,dp1,dp2,dp3,t]*Limit[dR2bt[p1,p2,p3,dp1,dp2,dp3,tt],tt->0]);
dABdBAt0[p1_,p2_,p3_,dp1_,dp2_,dp3_] := Limit[dABdBA[p1,p2,p3,dp1,dp2,dp3,t],t->0];
dABdBAt0[p1,p2,p3,dp1,dp2,dp3] - 2*(
    ff[p1,p2,p3,dp1,dp2,dp3]*R2bt[p1,p2,p3,dp1,dp2,dp3,0]
- 4*gg[p1,p2,p3,dp1,dp2,dp3]*R2at[p1,p2,p3,dp1,dp2,dp3,0]
) //Simplify
(******************************* should be 0 *******************************)


dR2t0[p1_,p2_,p3_,dp1_,dp2_,dp3_ ] := Limit[dR2t [p1,p2,p3,dp1,dp2,dp3,t],t->0];
dR2tx[p1_,p2_,p3_,dp1_,dp2_,dp3_ ] := dR2t0[p1,p2,p3,dp1,dp2,dp3] - dABdBAt0[p1,p2,p3,dp1,dp2,dp3] / R2bt[p1,p2,p3,dp1,dp2,dp3,0]^2 //Simplify;
(* dR2tx[p1,p2,p3,dp1,dp2,dp3] *)

xp1={1,2,3};
xp2={0,1,0};
xp3={1,0,0};
xdp1={1,2,3};
xdp2={-1,2,5};
xdp3={0,2,9};

dR2tx[xp1,xp2,xp3,xdp1,xdp2,xdp3]
(******************************* should be 0 *******************************)


dABdBAtTest1[p1_,p2_,p3_,dp1_,dp2_,dp3_] := 2*(    ff[p1,p2,p3,dp1,dp2,dp3]*R2bt[p1,p2,p3,dp1,dp2,dp3,0]
                                               - 4*gg[p1,p2,p3,dp1,dp2,dp3]*R2at[p1,p2,p3,dp1,dp2,dp3,0]
                                               ) //Simplify;

dR2t0[p1_,p2_,p3_,dp1_,dp2_,dp3_ ] := Limit[dR2t [p1,p2,p3,dp1,dp2,dp3,t],t->0];
dR2tx[p1_,p2_,p3_,dp1_,dp2_,dp3_ ] := dR2t0[p1,p2,p3,dp1,dp2,dp3] - dABdBAtTest1[p1,p2,p3,dp1,dp2,dp3] / R2bt[p1,p2,p3,dp1,dp2,dp3,0]^2 //Simplify;
(* dR2tx[p1,p2,p3,dp1,dp2,dp3] *)

xp1={1,2,3};
xp2={0,1,0};
xp3={1,0,0};
xdp1={1,2,3};
xdp2={-1,2,5};
xdp3={0,2,9};

dR2tx[xp1,xp2,xp3,xdp1,xdp2,xdp3]
(******************************* should be 0 *******************************)


dABdBAtTest2[p1_,p2_,p3_,dp1_,dp2_,dp3_] := 2*(    ff[p1,p2,p3,dp1,dp2,dp3]*R2b[p1,p2,p3]
                                               - 4*gg[p1,p2,p3,dp1,dp2,dp3]*R2a[p1,p2,p3]
                                              ) //Simplify;

dR2t0[p1_,p2_,p3_,dp1_,dp2_,dp3_ ] := Limit[dR2t [p1,p2,p3,dp1,dp2,dp3,t],t->0];
dR2tx[p1_,p2_,p3_,dp1_,dp2_,dp3_ ] := dR2t0[p1,p2,p3,dp1,dp2,dp3] - dABdBAtTest2[p1,p2,p3,dp1,dp2,dp3] / R2b[p1,p2,p3]^2 //Simplify;
(* dR2tx[p1,p2,p3,dp1,dp2,dp3] *)

xp1={1,2,3};
xp2={0,1,0};
xp3={1,0,0};
xdp1={1,2,3};
xdp2={-1,2,5};
xdp3={0,2,9};

dR2tx[xp1,xp2,xp3,xdp1,xdp2,xdp3]
(******************************* should be 0 *******************************)


dABdBAtTest2[p1_,p2_,p3_,dp1_,dp2_,dp3_] := 2*(    ff[p1,p2,p3,dp1,dp2,dp3]*R2b[p1,p2,p3]
                                                - 4*gg[p1,p2,p3,dp1,dp2,dp3]*R2a[p1,p2,p3]
                                               )                           / R2b[p1,p2,p3]^2 //Simplify;

dR2t0[p1_,p2_,p3_,dp1_,dp2_,dp3_ ] := Limit[dR2t [p1,p2,p3,dp1,dp2,dp3,t],t->0];
dR2tx[p1_,p2_,p3_,dp1_,dp2_,dp3_ ] := dR2t0[p1,p2,p3,dp1,dp2,dp3] - dABdBAtTest2[p1,p2,p3,dp1,dp2,dp3]  //Simplify;
(* dR2tx[p1,p2,p3,dp1,dp2,dp3] *)

xp1={1,2,4};
xp2={0,1,0};
xp3={1,0,0};
xdp1={1,2,3};
xdp2={-1,2,5};
xdp3={0,2,9};

dR2tx[xp1,xp2,xp3,xdp1,xdp2,xdp3]
(******************************* should be 0 *******************************)


R2[p1_,p2_,p3_] := ((p1-p2).(p1-p2)) * ((p2-p3).(p2-p3)) * ((p3-p1).(p3-p1)) / (4*Cross[p1-p2, p2-p3].Cross[p1-p2, p2-p3]);
R2a[p1_,p2_,p3_]:= ((p1-p2).(p1-p2)) * ((p2-p3).(p2-p3)) * ((p3-p1).(p3-p1));
R2b[p1_,p2_,p3_]:=                                                               (4*Cross[p1-p2, p2-p3].Cross[p1-p2, p2-p3]);
R2[p1_,p2_,p3_] := R2a[p1,p2,p3] / R2b[p1,p2,p3];
R2t [p1_,p2_,p3_,dp1_,dp2_,dp3_,t_] := R2 [p1+dp1*t,p2+dp2*t,p3+dp3*t];
dR2t [p1_,p2_,p3_,dp1_,dp2_,dp3_,t_] := D[R2t [p1,p2,p3,dp1,dp2,dp3,t],t];
(* R2 = R2a / R2b *)
f[a_,b_,c_,da_,db_,dc_] := ((da-db).(a-b)) * ((a-c).(a-c)) * ((b-c).(b-c));
ff[a_,b_,c_,da_,db_,dc_] := f[a,b,c,da,db,dc]+f[c,a,b,dc,da,db]+f[b,c,a,db,dc,da];
(* ff[...] = f(a,b,c) + f(c,a,b) + f(b,c,a) *)
gg[a_,b_,c_,da_,db_,dc_] := Cross[a-b,b-c] . (Cross[a-b, db-dc] + Cross[da-db, b-c]);
(* g[...] = (Pab x Pbc)^tr (Pab x Wbc + Wab x Pbc)
      Wbc = Wa - Wb = da - db
      Pab = Pa - Pb = a - b
      a x b = cross[a,b]
*)
dABdBAtTest2[p1_,p2_,p3_,dp1_,dp2_,dp3_] := 2*(    ff[p1,p2,p3,dp1,dp2,dp3]*R2b[p1,p2,p3]
                                                - 4*gg[p1,p2,p3,dp1,dp2,dp3]*R2a[p1,p2,p3]
                                               )                           / R2b[p1,p2,p3]^2 //Simplify;

dR2t0[p1_,p2_,p3_,dp1_,dp2_,dp3_ ] := Limit[dR2t [p1,p2,p3,dp1,dp2,dp3,t],t->0];
dR2tx[p1_,p2_,p3_,dp1_,dp2_,dp3_ ] := dR2t0[p1,p2,p3,dp1,dp2,dp3] - dABdBAtTest2[p1,p2,p3,dp1,dp2,dp3]  //Simplify;
(* dR2tx[p1,p2,p3,dp1,dp2,dp3] *)

xp1={1,2,4};
xp2={0,1,0};
xp3={1,0,0};
xdp1={1,2,3};
xdp2={-1,2,5};
xdp3={0,2,9};

dR2tx[xp1,xp2,xp3,xdp1,xdp2,xdp3]
(******************************* should be 0 *******************************)



















p1={1,2,4};
p2={0,1,0};
p3={1,0,0};
dp1={1,2,3};
dp2={-1,2,5};
dp3={0,2,9};



P12dist2 = (p1-p2).(p1-p2);
P23dist2 = (p2-p3).(p2-p3);
P31dist2 = (p3-p1).(p3-p1);
Dotp12dp12 = (dp1-dp2).(p1-p2);
Dotp23dp31 = (dp2-dp3).(p2-p3);
Dotp31dp31 = (dp1-dp3).(p1-p3);
DotCroP12P23CroP12Dp23CroDp12P23 = Cross[p1-p2, p2-p3] . (Cross[p1-p2, dp2-dp3] + Cross[dp1-dp2, p2-p3]);
(* area = 0.5 |AB x AC| *)
Div2 = Cross[p1-p2, p2-p3].Cross[p1-p2, p2-p3];
Div4 = Div2^2;


dR2dt = 2*(+ ( Dotp31dp31 * P12dist2 * P23dist2
             + Dotp23dp31 * P12dist2 * P31dist2
             + Dotp12dp12 * P23dist2 * P31dist2 )/(4 * Div2)
           - ( DotCroP12P23CroP12Dp23CroDp12P23 * P12dist2 * P23dist2 * P31dist2)/(4 * Div4)
          )
(* notation in implementation *)
(******************************* all values should be the same *******************************)
dR2t0[p1,p2,p3,dp1,dp2,dp3]
(******************************* all values should be the same *******************************)
dABdBAtTest2[p1,p2,p3,dp1,dp2,dp3]
(******************************* all values should be the same *******************************)


R0 = 0.5*Sqrt[P12dist2 * P23dist2 * P31dist2 / Div2];
dRdR2 = 1/(2* R0);
dR2dtTest1 = dRdR2 * dR2dt
(******************************* all values should be the same *******************************)
dR2dtTest2 = (( Dotp31dp31*P12dist2*P23dist2 + Dotp23dp31*P12dist2*P31dist2 + Dotp12dp12*P23dist2*P31dist2 )/(4 * R0 * Div2)
        -( DotCroP12P23CroP12Dp23CroDp12P23 * R0 )/(1 * Div2)
          )
(* notation in the paper *)
(******************************* all values should be the same *******************************)
dR2dtTest1 - dR2dtTest2
(******************************* should be 0 *******************************)



